Let configure generate stdlib/sys.ml

shindere opened the pull request:
This PR follows the discussion on #1057.

It contains just the bit that transforms `stdlib/sys.ml` into a proper
`configured module`, so to speak, in the sense that it is generated
by the approrpiate autoconf machinery rather than being generated
by a sed invocation during the build itself.


With this commit, the stdlib/sys.ml file is generated at configure time
rather than by stdlib/Makefile.

@nojb would you mind reviewing this, please?

nojb reviewed (approved):
LGTM

shindere commented:
Nicolás Ojeda Bär (2021/08/23 06:48 -0700):
> @nojb approved this pull request.
>
> LGTM

Thanks!

If CI passes and nobody expresses dissatisfaction meanwhile, I'll merge.

xavierleroy commented:
Appveyor fails during bootstrap.  Apparently, a `make -C stdlib clean` in the middle of bootstrap causes `stdlib/sys.ml` to be removed.

shindere committed:
Let configure generate stdlib/sys.ml

With this commit, the stdlib/sys.ml file is generated at configure time
rather than by stdlib/Makefile.

shindere commented:
Many thanks for the detailed explanation, Xavier!

I did add a distclean target to remove `sys.ml` but forgot to remove the
removal from the clean target. I just pushed a revised version, I hope
it will make CI happy.

gadmm commented:
Can you point us to where this change was previously discussed? The link you gave seems to lead in the wrong place.

Not that I feel very involved in the matter but the change surprises me. I have already modified `sys.mlp` in the past (none in a way that ended up in one of my PRs), and I could easily imagine losing a day's work (exaggerating things a bit) over not expecting that one need to re-run configure to re-build `sys.ml` instead of having it done automatically by the build system. (Not counting the times I omit to re-run configure when switching branches, but I don't know if other people are lazy like me.)

gadmm commented on stdlib/sys.ml.in:
```diff
@@ -14,10 +15,6 @@
 (**************************************************************************)
 
-(* WARNING: sys.ml is generated from sys.mlp.  DO NOT EDIT sys.ml or
-   your changes will be lost.
```
If this change goes through, it would be nice if there still was a similar warning (mentioning `sys.ml.in` and `configure`).

shindere commented:
Guillaume Munch-Maccagnoni (2021/08/23 18:18 -0700):
> Can you point us to where this change was previously discussed? The
> link you gave seems to lead in the wrong place.

I didn't even remmber giving a link. It was discussed in thecourse of
the discussion on #10576, look for @nobj's comment there.

> Not that I feel very involved in the matter but the change surprises
> me. I have already modified `sys.mlp` in the past (none in a way that
> ended up in one of my PRs), and I could easily imagine losing a day's
> work (exaggerating things a bit) over not expecting that one need to
> re-run configure to re-build `sys.ml` instead of having it done
> automatically by the build system. (Not counting the times I omit to
> re-run configure when switching branches, but I don't know if other
> people are lazy like me.)

When configure generates `foo` from `foo.in`, provided that you add the
***@***.***_input@` macro invocation to `foo.in` then the macro call
will be expanded in `foo` into a text saying that `foo` has been
generated from `foo.in` by configure.

This has been done in this way in this PR.

shindere commented:
Guillaume Munch-Maccagnoni (2021/08/23 18:24 -0700):
> @gadmm commented on this pull request.
>
>
>
> > @@ -14,10 +15,6 @@
>  (*                                                                        *)
>  (**************************************************************************)
>
> -(* WARNING: sys.ml is generated from sys.mlp.  DO NOT EDIT sys.ml or
> -   your changes will be lost.
>
> If this change goes through, it would be nice if there still was a
> similar warning (mentioning `sys.ml.in` and `configure`).

See my previous response.

May I suggest that you just try the branch locally and experiment how
things work?

gadmm commented:
It is just that since you mention that this was already discussed, and at the same time you ask if anyone expresses dissatisfaction, I thought you might want people to have the occasion to get informed about the discussion before you merge. I have now read the comment you were referring to and I confirm that this PR surprises me.

(Now the scenario I described will probably not happen in this way, but my reaction comes from my positive experience when working on a large code base where incremental builds worked well.)

xavierleroy commented:
Let's not fight about this change.  Many projects use configure scripts to generate file F from file F.in, so it should not be too surprising that configure needs to be re-run if stdlib/sys.ml.in changes.

shindere commented:
Guillaume Munch-Maccagnoni (2021/08/24 16:26 +0000):
> It is just that since you mention that this was already discussed, and at the same time you ask if anyone expresses dissatisfaction, I thought you might want people to have the occasion to get informed about the discussion before you merge. I have now read the comment you were referring to and I confirm that this PR surprises me.
>
> (Now the scenario I described will probably not happen in this way,
> but my reaction comes from my positive experience when working on a
> large code base where incremental builds worked well.)

Yes, I understand your concern. It's just that our build system is full
of calls to scripts, which make it slower on Windows, which @dra27 will
ocnfirm.

In the long run, I do hope we will move to a dynamic computation of
dependencies, which should fix the incremental builds.

gadmm commented:
Thanks for your explanation!

dra27 commented:
Regarding the concern about editing `stdlib/sys.ml.in`, we could do:

```diff
diff --git a/stdlib/Makefile b/stdlib/Makefile
index df34bc2d49..92cc615be3 100644
--- a/stdlib/Makefile
+++ b/stdlib/Makefile
@@ -200,6 +200,11 @@ stdlib.cma: $(OBJS)
 stdlib.cmxa: $(OBJS:.cmo=.cmx)
        $(CAMLOPT) -a -o $@ $^

+sys.ml: sys.ml.in
+       @echo 'stdlib/sys.ml.in is newer than stdlib/sys.ml - you most likely'; \
+   echo 'need to re-run ./config.status (or touch stdlib/sys.ml)'
+       @false
+
 .PHONY: distclean
 distclean: clean
        rm -f sys.ml
```

which causes an error to be displayed by `make` if you edit `stdlib/sys.ml.in` and don't re-run `configure` (or, as the message says, re-run `./config.status` which is faster).

Happy to flip that into a PR. In theory, `stdlib/Makefile` could just call `../config.status` but that regenerates other files as well, so I think it's better to be explicit about the need to run the command.

shindere commented:
I'm not found of the idea, I have to say. If we do it here then we
should do it for s.h, m.h and maybe other files. Ther's the comment at
thetop of the files telling people they are generated already.
